### ISSUE TITLE
Confusing Naming of Component State Enums: `ComponentStatus` and `ComponentStateCode`

### ISSUE THREAD

**Issue by tiyash-basu-frequenz** [frequenz-floss/frequenz-api-common#264]
### What's needed?

We currently have two component state enums:

- [`ComponentStatus`](https://github.com/frequenz-floss/frequenz-api-common/blob/3d0aa691af167f135801ffe58584612127ece6d3/proto/frequenz/api/common/v1/microgrid/components/components.proto#L101-L117): This enum depicts the state of our interaction with a component (e.g., whether we can read from or write to it).
- [`ComponentStateCode`](https://github.com/frequenz-floss/frequenz-api-common/blob/3d0aa691af167f135801ffe58584612127ece6d3/proto/frequenz/api/common/v1/microgrid/components/components.proto#L259C6-L259C24): This enum represents the actual state of the component as derived from the component directly.

The current naming can potentially lead to confusion between `ComponentStatus` and `ComponentStateCode`. They could be mistaken for referring to the same entity. Furthermore, `ComponentStatus` lacks specificity, as it seems to indicate the component's overall status rather than focusing on the aspect of interaction. 

We need to address the naming issue to improve readability.

### Proposed solution

<details>
<summary>Proposal 1</summary>
<br>
We can rename the `ComponentStatus` enum to something more appropriate.

One proposal is `ComponentAccessLevel`. This correctly depicts the purpose of this entity.
</details>

<details>
<summary>Proposal 2</summary>
<br>

We flip the concept, and do the following:

**`Active` mode**

As soon as a component moves to this mode, we try to start it up. Note that this cannot be guaranteed to succeed, since a component could refuse to start up, e.g., due to the component's internal configuration.

In this mode, clients are allowed to read data from the component, and send commands to it. Sending commands to a component in this mode can be prevented by setting its respective bounds to (0,0), which would make the component read-only.

**`Disabled` mode**

As soon as a component moves to this mode, we try to stop it. Again, this cannot be guaranteed to succeed due the same reasons mentioned above.

In this mode, clients do not receive any data from the component, and cannot send any control commands to it.

In this concept, `RequestedComponentMode` may be a better name.

</details>

<details>
<summary>Proposal 3</summary>
<br>

This extends Proposal 2, with an intermediate mode for switching the component off while still getting data from it.

**`Active` mode**

As soon as a component moves to this mode, we try to start it up. Note that this cannot be guaranteed to succeed, since a component could refuse to start up, e.g., due to the component's internal configuration.

In this mode, clients are allowed to read data from the component, and send commands to it. Sending commands to a component in this mode can be prevented by setting its respective bounds to (0,0), which would make the component read-only.

**`Off` mode**

As soon as a component moves to this mode, we try to stop it. Again, this cannot be guaranteed to succeed due the same reasons mentioned above.

In this mode, clients keep receiving data from the component, but cannot send any control commands to it. This mode allows users to receive a feedback that their action to move the component to mode `Off` has worked or not.

**`Inaccessible` mode**

As soon as a component moves to this mode, we assume that the component is not available on the network, and we do not attempt to read from it.

In this mode, clients do not receive any data from the component, and cannot send any control commands to it.

</details>

#### Proposal 4

**`Enabled` mode**

As soon as a component moves to this mode, we try to start it up. (Note that this cannot be guaranteed to succeed, since a component could refuse to start up, e.g., due to the component's internal configuration.)

In this mode, clients are allowed to read data from the component, and send commands to it. Sending commands to a component in this mode can be prevented by setting its respective bounds to (0,0), which would make the component read-only.

**`Disabled` mode**

As soon as a component moves to this mode, we try to stop it. (Again, this cannot be guaranteed to succeed due the same reasons mentioned above.)

In this mode,
- we will try to disconnect the component from the microgrid, by opening its relays.
- we will prevent applications from sending control commands to the component.
- we will keep getting data from this component.

### Use cases

- switch a component on - e.g., closing its AC/DC relays
- switch a component off - e.g., opening its AC/DC relays. Note that depending upon the component model, it may still be able to stay online and allow us to read from it.
- ~run maintenance activities on components  - e.g., battery calibration, firmware updates.~
- add a placeholder component to the component graph, to indicate its existence in case we cannot reach it over the network due to hardware-installation-related challenges.

**Comment by llucax** [comment]
Really, is `ComponentStatus` supposed to be about access level? Currently is only `ACTIVE` and `INACTIVE`, from the current values maybe `ComponentOperationalStatus` might be more precise, or we could have it as a `bool` if being active or inactive is independent from the *access level*.

**Comment by tiyash-basu-frequenz** [comment]
As of now, it is in fact about access - whether we should read from (and write to) a component or not. In practice, this could be used for a component that is electrically connected to a microgrid, but we do not have physical access to it. This has nothing to do with the component itself being active or inactive.

I do not think `ComponentOperationalStatus` helps. It is closer in meaning to the actual state of the component.
This would mean that `ACTIVE` and `INACTIVE` are not appropriately named either.

Or maybe we flip the concept, and say that if a component is moved to an `INACTIVE`, then we stop the component, and prevent clients from reading from it or writing to it. This approach of course has its own naming challenges.

**Comment by tiyash-basu-frequenz** [comment]
I have added a 2nd proposed solution, and am inclined towards it.

**Comment by thomas-nicolai-frequenz** [comment]
> As soon as a component moves to this state, we try to start it up. Note that this cannot be guaranteed to succeed, since a component could refuse to start up, e.g., due to the component's internal configuration.

I think we need to be really clear on this that disabled would mean that a Microgrid API client can't use the component anymore but that it does not mean its really shut down. I feel thats something that would need clarification somehow.

**Comment by llucax** [comment]
Interpretation of proposal 3 from the client (app) side:

* `Active` mode: we put the component in the component graph and use it to come up with formulas, we try to read from it and to control it.
* `Off` mode: we put the component in the component graph and take it into account when building formulas. We try to read from it but we don't try to control it.
* `Inaccessible`: we put the component in the component graph and take it into account when building formulas, we want to read from it, but we need to find a proxy to read from it. We don't try to control it.

@shsms Maybe you can double-check if the above makes sense or am I missing anything?

**Comment by tiyash-basu-frequenz** [comment]
Right. (I just added a proposal 3 to the description, after a discussion)

**Comment by llucax** [comment]
> **`Inaccessible` mode**
> 
> As soon as a component moves to this state, we try to stop it. Again, this cannot be guaranteed to succeed due the same reasons mentioned above.

Why would you try to stop it in this case? If we know the component is **inaccessible** we can't contact it, isn't it this the whole point of having a mode for inaccessible components?

**Comment by llucax** [comment]
(I took the liberty of editing the issue description to replace all occurrences of **state** with **mode** to make it more clear).

**Comment by tiyash-basu-frequenz** [comment]
> > **`Inaccessible` mode**
> > As soon as a component moves to this state, we try to stop it. Again, this cannot be guaranteed to succeed due the same reasons mentioned above.
> 
> Why would you try to stop it in this case? If we know the component is **inaccessible** we can't contact it, isn't it this the whole point of having a mode for inaccessible components?

oops, copy-paste error. Fixed.

**Comment by llucax** [comment]
The naming should probably a different discussion, and for the purpose of the discussion here I think we should keep using `Active`, `Off`, and `Inaccessible`, but I wanted to point out before I forget that `Off` seems a bit... off (🤡) if it still can send data. I wonder if `Disabled` or `Inactive` could be better for this, as it is not doing actual useful work (and also the opposite of `Active`), but it is not necessarily completely *off* (as in shut down, completely unresponsive).

**Comment by ktickner** [comment]
> Note that this cannot be guaranteed to succeed, since a component could refuse to start up, e.g., due to the component's internal configuration.

How does the user gain feedback whether or not switching mode has succeeded?

> Inaccessible

I don't like this naming convention. It sounds like something the user has no control over, i.e. it is not a state initiated by the user. These are all actions the user can take through the UI, so they should be reflective of the fact that they are controllable, and should be mutually controllable (i.e. the user PUTS the component in this mode, and can take it out of this mode again).
If inaccessible means the user can't change the mode, or is a reflection of some state that the user has no control over, it should be derived rather than relying on user input as this tells the user they can take the component out of this state again through the UI - does this make sense?

**Comment by shsms** [comment]
> Interpretation of proposal 3 from the client (app) side:
> 
>    * `Active` mode: we put the component in the component graph and use it to come up with formulas, we try to read from it and to control it.
> 
>    * `Off` mode: we put the component in the component graph and take it into account when building formulas. We try to read from it but we don't try to control it.
> 
>    * `Inaccessible`: we put the component in the component graph and take it into account when building formulas, we want to read from it, but we need to find a proxy to read from it. We don't try to control it.

Yup, all that makes sense.  We still have to decide how to distribute for example PV power at a location where there are some `inaccesible` PV inverters and some `active`.  But that's a separate discussion.

**Comment by llucax** [comment]
As an update, in the latest meeting where we discussed this, we agreed to only have 2 modes: `Enabled` and `Disabled`. If a component is inaccessible, we will handle it as any other error case where a component is not working, and we need to find fallbacks or proxies to interact with it (like an inverter or meter). The only requirement is that the component that is not supported/inaccessible must be present in the component graph, then the regular failure logic should take over.

**Comment by tiyash-basu-frequenz** [comment]
Added a 4th proposal, with API/server-side requirements, from the call Luca mentioned above.

**Comment by ktickner** [comment]
> (Note that this cannot be guaranteed to succeed, since a component could refuse to start up, e.g., due to the component's internal configuration.)

Please remind me what happens if the start/stop operation does not succeed? @tiyash-basu-frequenz

**Comment by tiyash-basu-frequenz** [comment]
> > (Note that this cannot be guaranteed to succeed, since a component could refuse to start up, e.g., due to the component's internal configuration.)
> 
> Please remind me what happens if the start/stop operation does not succeed? [@tiyash-basu-frequenz](https://github.com/tiyash-basu-frequenz)

Then nothing visible happens to the component, and it keeps behaving as if it never got the request to start/stop.

**Comment by ktickner** [comment]
> Then nothing visible happens to the component, and it keeps behaving as if it never got the request to start/stop.

But there ought to be feedback on the UI, no? Or does the UI just pretend that the operation was successful and we trust the user to know the difference?

**Comment by tiyash-basu-frequenz** [comment]
Addressed by #307
